astropy/wcs/src/wcslib_celprm_wrap.c

```diff
@@ -0,0 +1,486 @@
+#define NO_IMPORT_ARRAY
+
+#include "astropy_wcs/wcslib_celprm_wrap.h"
+#include "astropy_wcs/wcslib_prjprm_wrap.h"
+
+#include <wcs.h>
+#include <wcsprintf.h>
+#include <cel.h>
+#include <prj.h>
+#include <wcserr.h>
+#include <numpy/npy_math.h>
+
+#include <stdio.h>
+
+/*
+ It gets to be really tedious to type long docstrings in ANSI C syntax
+ (since multi-line strings literals are not valid).  Therefore, the
+ docstrings are written in doc/docstrings.py, which are then converted
+ by setup.py into docstrings.h, which we include here.
+*/
+#include "astropy_wcs/docstrings.h"
+#include "astropy_wcs/wcslib_wrap.h"
+
+
+PyObject** cel_errexc[7];
+
+
+static int wcslib_cel_to_python_exc(int status)
+{
+    if (status > 0 && status < 7) {
+        PyErr_SetString(*cel_errexc[status], cel_errmsg[status]);
+    } else if (status > 6) {
+        PyErr_SetString(
+            PyExc_RuntimeError,
+            "Unknown WCSLIB celprm-related error occurred.");
+    }
+    return status;
+}
+
+
+static int is_readonly(PyCelprm* self)
+{
+    if (self != NULL && self->owner != NULL) {
+        PyErr_SetString(
+                PyExc_AttributeError,
+                "Attribute 'cel' of 'astropy.wcs.Wcsprm' objects is read-only.");
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+
+static int is_cel_null(PyCelprm* self)
+{
+    if (self->x == NULL) {
+        PyErr_SetString(
+                PyExc_MemoryError,
+                "Underlying 'celprm' object is NULL.");
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+
+/***************************************************************************
+ * PyCelprm methods                                                        *
+ ***************************************************************************/
+
+static PyObject* PyCelprm_new(PyTypeObject* type, PyObject* args, PyObject* kwds)
+{
+    PyCelprm* self;
+    self = (PyCelprm*)type->tp_alloc(type, 0);
+    if (self == NULL) return NULL;
+    self->owner = NULL;
+    self->prefcount = NULL;
+
+    if ((self->x = calloc(1, sizeof(struct celprm))) == 0x0) {
+        PyErr_SetString(PyExc_MemoryError,
+        "Could not allocate memory for celprm structure.");
+        return NULL;
+    }
+    if ((self->prefcount = (int*) malloc(sizeof(int))) == 0x0) {
+        PyErr_SetString(PyExc_MemoryError, "Could not allocate memory.");
+        free(self->x);
+        return NULL;
+    }
+
+    if (wcslib_cel_to_python_exc(celini(self->x))) {
+        free(self->x);
+        free(self->prefcount);
+        return NULL;
+    }
+    *(self->prefcount) = 1;
+    return (PyObject*)self;
+}
+
+
+static int PyCelprm_traverse(PyCelprm* self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->owner);
+    return 0;
+}
+
+
+static int PyCelprm_clear(PyCelprm* self)
```

Comment:
From https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_clear:

> ... there’s rarely a good reason to avoid implementing tp_clear.

Does it hurt to have it?

Reply:
Only in that extraneous code is more stuff someone has to grok later if they want to understand this file.  It's not a big deal either way.